21.04.2020

VbKoshak opened the pull request.

VbKoshak committed:
attemp#1

VbKoshak committed:
update gitignore

VbKoshak committed:
.gitignore fix

shukal94 commented on src/Campaigns.java:
```diff
@@ -0,0 +1,5 @@
+public class Campaigns {
```
Сampaign

shukal94 replied:
why did you mark it as resolved?

shukal94 commented on src/Campaigns.java:
```diff
@@ -0,0 +1,5 @@
+public class Campaigns {
+  private String name;
```
you can add start and end date

shukal94 replied:
same as above

shukal94 commented on src/Cart.java:
```diff
@@ -0,0 +1,6 @@
+
+public class Cart {
+  private HashMap<Long,String> purchases;
+  private long totalPrice;
```
subtotal price may be
so your total price = subtotal + tax (at the order) + shipment fee (can be assigned as shipment method at the order)

shukal94 replied:
where is update?

shukal94 commented on src/ProductAttribute.java:
```diff
@@ -0,0 +1,4 @@
+public class ProductAttribute {
```
let's mark this class as abstract, and derive attributes as concrete classes. like Size can be a String, Color can be a int, Images is a String or base64, etc.

VbKoshak replied:
As i planned, size color ets would be objects of class ProductAttribute.
So, color object would be like
Productattribute color{
    name: "color",
    values: ["black","white","green"]
}

shukal94 replied:
color is good to be int (hex code) as we can assign this code in css on front-end
and if I want to expand the nuber of attributes you may need one big class that know everything about attributes, it's better to make every class single-responsible

shukal94 replied:
color is more convenient in int (hex value converted into decimal)

shukal94 replied:
and do not try to make one universal class, it is better for readability and debugging to make a little single-responsible classes

shukal94 commented on src/User.java:
```diff
@@ -0,0 +1,9 @@
+  private String login;
+  private String passwordHash;
+  private UserOrderInfo[] orderInfo;
+  private Order[] OrderHistory;
```
duplicate

shukal94 replied:
where is update?

shukal94 commented on task2.iml:
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
include this file into gitignore

shukal94 reviewed (changes requested):
1) please, find out some gitignore for idea
2) define getters/setters and constructors as we may expand this code in future tasks
3) collaborate with someone from group about the common comments

good, but there are some points to improve

VbKoshak committed:
gitignore update

VbKoshak committed:
iml file rename

VbKoshak committed:
delete iml file

VbKoshak commented:
sry for missunderstanding, i have marked resolved problems that got resolved locally. i didn't push yet

P.S. will not do the same in future

VbKoshak commented:
as there are such a small issues as grammar mistakes, decided that i will not take futher conversation after a fix

VbKoshak commented:
one more question: is it necessary to make good commits for now? (like 1 feature/fix/... 1 commit) or it doesn't matter right now?

shukal94 commented:
make sense

shukal94 commented on src/Cart.java:
```diff
@@ -0,0 +1,6 @@
+
+public class Cart {
+  private HashMap<Long,String> purchases;
+  private long totalPrice;
```
its better to store Product as a value in purchases
private long totalPrice; - subtotal

shukal94 commented on src/PriceBook.java:
```diff
@@ -0,0 +1,6 @@
+
+public class PriceBook {
+  // all attributes going through hash function to get HashString (String) that will lead to price (long)
+  private Hashtable<String, Long> book  = new Hashtable<>();
```
let's store a product as a value

VbKoshak replied:
Maybe hashString can already hash the Product and attributes so there Will be common Pricebook for all goods our Shop sells?

shukal94 replied:
You have to map that hash and proper product on that case

shukal94 commented on src/Product.java:
```diff
@@ -0,0 +1,12 @@
+  private long id;
+  private String name;
+  private ProductAttribute[] attributes;
+  private PriceBook priceBook;
```
we don't have to know about the price

shukal94 commented on src/Product.java:
```diff
@@ -0,0 +1,12 @@
+  private ProductAttribute[] attributes;
+  private PriceBook priceBook;
+  private String description;
+  private HashMap<String, String> pictures = new HashMap<>(); //attributeHash(String), link to photo (string)
```
pictures are attribute

shukal94 commented on src/Promotion.java:
```diff
@@ -0,0 +1,5 @@
+public class Promotion extends PromotionRule{
+  private String name;
+  private String description;
+  private String[] requirements;
```
one single req will be enough

shukal94 commented on src/User.java:
```diff
@@ -0,0 +1,9 @@
+public class User extends Guest{
```
why user extends guest?

shukal94 replied:
guest IS A user

VbKoshak replied:
So, can i rename it as an authorizedUser? Because any1 Who visits the Shop is already guest and there is no point in creating parent class to it as for me.

shukal94 replied:
I meant that no need to extend from user. Make user abstract and derive all kind of users you want

VbKoshak commented:
So what about rules in comments, as i see some of them. What's the rule you use in mkst projects?

shukal94 commented:
If you have a question comment, if no, just push

VbKoshak committed:
fix#1

add getters/setters
lexical grammar fix
duplicate fix
add start and end date to campaign
add subtotal and total price of order

VbKoshak committed:
java-core-2: fix#2

add abstract user
class Image
ProductAttribute -> abstract
PriceBook belongs to ProductCatalog
images in Product now contains Image instead of link
PriceBook hash from String to Int

shukal94 commented on src/Guest.java:
```diff
@@ -0,0 +1,10 @@
+    //..
+  }
+
+  public Guest{
```
()

shukal94 commented on src/Image.java:
```diff
@@ -0,0 +1,17 @@
+public class Image extends ProductAttribute {
```
Images maybe

VbKoshak replied:
It is one photo, not an array. Why should it be Images?

shukal94 commented on src/Order.java:
```diff
@@ -0,0 +1,99 @@
+    waitingPayment,
+    delivering
+  }
+
```
please move class fields to the top

shukal94 reviewed (commented):
lgtm

VbKoshak committed:
java-core-2: fix#3

small fixes
lexical mistakes
clean code
add Cart constructor
